### PR TITLE
[Bug fix] better GPU flash attention compatibility

### PR DESCRIPTION
**Pull request by kelvin-zou**
Fixing a few small issues on flash attention:
1. fix fp32 support, fall back to triton kernel due to lack of cudnn support (we use it in eval in most cases), also fixed a SRAM OOM issue in triton kernel bwd pass on fp32.
2. fix bias support, fall back to triton kernel due to limited support in cudnn (may encounter OOM in bwd pass in cudnn). 
3. Added bunch of tests to verify both fixes.

### CODE REVIEW

**Review by markblee — COMMENTED** [review]
Thanks @kelvin-zou -- can you take a look at the internal comment?

**Comment by kelvin-zou** [comment]
> Thanks @kelvin-zou -- can you take a look at the internal comment?

Did and already replied there. Also changelog updated.

**Review by ruomingp — APPROVED** [review]

**Comment by ruomingp** [comment]
> > Thanks @kelvin-zou -- can you take a look at the internal comment?
> 
> Did and already replied there. Also changelog updated.

Is this resolved?

File: axlearn/common/flash_attention/gpu_attention.py [hunk 1]
```diff
@@ -590,6 +590,9 @@ def _mha_backward(
 
     # NOTE: temporarily removed the "xla" branch, which seems unused.
     if backward_pass_impl == "triton":
+        # We must shrink the block size for float32 inputs to avoid OOM during bwd pass.
+        if jnp.float32 in (q.dtype, k.dtype, v.dtype):
+            block_q = block_k = 64
```
Comment (samos123) [inline axlearn/common/flash_attention/gpu_attention.py:595]: Should we use min instead, in the case of block_q and block_k being less than than 64?
```
block_q = min(64, block_q)
block_k = min(64, block_k)
```

Not sure if there is any scenario where block_k and block_q are less than 64 though.

Comment (samos123) [reply axlearn/common/flash_attention/gpu_attention.py:595]: looks like flash attention min is set to 64. So resolving this comment.
image

**Comment by kelvin-zou** [comment]
> > > Thanks @kelvin-zou -- can you take a look at the internal comment?
> > 
> > 
> > Did and already replied there. Also changelog updated.
> 
> Is this resolved?

Yes, already added the integration test in the internal PR.

**Review by markblee — APPROVED** [review]

**Pull request merged**